archspec/cpu/cpuid.py

```diff
@@ -0,0 +1,697 @@
+            old_protect = ctypes.c_ulong(0)
+            pfnVirtualProtect = ctypes.windll.kernel32.VirtualProtect
+            res = pfnVirtualProtect(
+                ctypes.c_void_p(self.address),
+                ctypes.c_size_t(size),
+                PAGE_EXECUTE,
+                ctypes.byref(old_protect),
+            )
+            if not res:
+                raise Exception("Failed VirtualProtect")
+
+            # Flush Instruction Cache
+            # First, get process Handle
+            if not self.prochandle:
+                pfnGetCurrentProcess = ctypes.windll.kernel32.GetCurrentProcess
+                pfnGetCurrentProcess.restype = ctypes.c_void_p
+                self.prochandle = ctypes.c_void_p(pfnGetCurrentProcess())
+            # Actually flush cache
+            res = ctypes.windll.kernel32.FlushInstructionCache(
+                self.prochandle, ctypes.c_void_p(self.address), ctypes.c_size_t(size)
+            )
+            if not res:
+                raise Exception("Failed FlushInstructionCache")
+        else:
+            from mmap import (
+                mmap,
+                MAP_PRIVATE,
+                MAP_ANONYMOUS,
+                PROT_WRITE,
+                PROT_READ,
+                PROT_EXEC,
+            )
+
+            # Allocate a private and executable memory segment the size of the machine code
+            machine_code = bytes.join(b"", self.machine_code)
+            self.size = len(machine_code)
+            self.mm = mmap(
+                -1,
+                self.size,
+                flags=MAP_PRIVATE | MAP_ANONYMOUS,
+                prot=PROT_WRITE | PROT_READ | PROT_EXEC,
+            )
+
+            # Copy the machine code into the memory segment
+            self.mm.write(machine_code)
+            self.address = ctypes.addressof(ctypes.c_int.from_buffer(self.mm))
+
+        # Cast the memory segment into a function
+        functype = ctypes.CFUNCTYPE(self.restype, *self.argtypes)
+        self.func = functype(self.address)
+
+    def run(self):
+        # Call the machine code like a function
+        retval = self.func()
+
+        return retval
+
+    def free(self):
+        # Free the function memory segment
+        if is_windows:
+            MEM_RELEASE = ctypes.c_ulong(0x8000)
+            ctypes.windll.kernel32.VirtualFree(
+                ctypes.c_void_p(self.address), ctypes.c_size_t(0), MEM_RELEASE
+            )
+        else:
+            self.mm.close()
+
+        self.prochandle = None
+        self.mm = None
+        self.func = None
+        self.address = None
+        self.size = 0
+
+
+class CPUID:
+    def _asm_func(self, restype=None, argtypes=(), machine_code=[]):
+        asm = ASM(restype, argtypes, machine_code)
+        asm.compile()
+        return asm
+
+    def _run_asm(self, *machine_code):
+        asm = ASM(ctypes.c_uint32, (), machine_code)
+        asm.compile()
+        retval = asm.run()
+        asm.free()
+        return retval
```

Comment:
Do we really need to run assembly code? Isn't there any kind of simple inspection that we can do on Windows? @johnwparent

Reply:
@isuruf is `wmic cpu list` not sufficient? What detail is lacking from that command? Additionally, there are other methods to query the registry, i.e. `winreg`, would that work?

Reply:
@isuruf to save you some time, this is what `wmic cpu list /format:list` gives me for my machine:
```
AddressWidth=64
Architecture=9
Availability=3
Caption=Intel64 Family 6 Model 158 Stepping 13
ConfigManagerErrorCode=
ConfigManagerUserConfig=
CpuStatus=1
CreationClassName=Win32_Processor
CurrentClockSpeed=2304
CurrentVoltage=10
DataWidth=64
Description=Intel64 Family 6 Model 158 Stepping 13
DeviceID=CPU0
ErrorCleared=
ErrorDescription=
ExtClock=100
Family=207
InstallDate=
L2CacheSize=2048
L2CacheSpeed=
LastErrorCode=
Level=6
LoadPercentage=8
Manufacturer=GenuineIntel
MaxClockSpeed=2304
Name=Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
OtherFamilyDescription=
PNPDeviceID=
PowerManagementCapabilities=
PowerManagementSupported=FALSE
ProcessorId=BFEBFBFF000906ED
ProcessorType=3
Revision=
Role=CPU
SocketDesignation=U3E1
Status=OK
StatusInfo=3
Stepping=
SystemCreationClassName=Win32_ComputerSystem
SystemName=DVORAK
UniqueId=
UpgradeMethod=52
Version=
VoltageCaps=
```

Reply:
@johnwparent, processorid is the only information there which can be used for feature detection. It contains EAX, EDX values and model numbers. We need `EBX` and `ECX` values from cpuid for detection of things like avx2, avx512.
The model number is useful, but then we would need to carry a database of model numbers to features which we would have to update for each model that each hardware vendor produces and is not feasible.

Reply:
`wmic` suggestion aside, after more investigation I see no other way to get cpu flags, seems to be the only way is C preproc/C program/Assembly via `CPUID` which is how `/proc/cpuinfo` is created anyway.

Reply:
@isuruf For your use case, do you need feature parity with `linux` on detection, or detecting x86_64 levels would suffice?

Reply:
Yeah, the levels would suffice, but that API is limited and we can't distinguish Knight's landing vs x86_64 level 4 since there's no way to check for `avx512vl` and `avx512dq` if I'm reading the docs correctly.

Reply:
This seems interesting: https://github.com/albertony/cpufeatures?tab=readme-ov-file#cpufeatures-1

Reply:
This snippet is interesting too: https://learn.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=msvc-170

Reply:
There's also `!cpuinfo` which is the assembler instruction that provides a lot of this information.

Reply:
> This seems interesting: https://github.com/albertony/cpufeatures?tab=readme-ov-file#cpufeatures-1

> This snippet is interesting too: https://learn.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=msvc-170

Both of these would require compiling a C file which would complicate things. This basically runs the same ASM as is done by py-cpuinfo.

> There's also !cpuinfo which is the assembler instruction that provides a lot of this information.

You mean `cpuid`? That's what `py-cpuinfo` and others use.

Reply:
Sorry, yes I did mean to write `cpuid`, but I did also mean to include `!cpuinfo`, which is an extension for the kernel debugger, just mashed the thoughts into one sentence.

Reply:
@alalazo, outside of the significant amount of `py-cpuinfo` code we're vendoring here (I think we both agree that we should roll our own approach here) is the concern with the compiler/assembler the fragility or the performance (or perhaps both)?